IdentityCore/src/util/NSString+MSIDExtensions.m

```diff
@@ -71,208 +38,45 @@ @implementation NSString (MSIDExtensions)
-            pbDecoded[ib++] = (b1 << 4) | (b2 >> 2);
-            
-            if (ib < cbDecodedSize) {
-                pbDecoded[ib++] = (b2 << 6) | b3;
-            }
-        }
-    }
-    
-    // Place the result in a NSData object and then free it.
-    NSData *result = [NSData dataWithBytes:pbDecoded length:cbDecodedSize];
-    
-    free( pbDecoded );
-    
-    return result;
+    NSUInteger paddedLength = encodedString.length + (4 - (encodedString.length % 4));
+    NSString *paddedString = [encodedString stringByPaddingToLength:paddedLength withString:@"=" startingAtIndex:0];
+    NSData *data = [[NSData alloc] initWithBase64EncodedString:paddedString options:NSDataBase64DecodingIgnoreUnknownCharacters];
+    return data;
 }
 
-- (NSString *)msidBase64UrlDecode
+
+// Base64 URL encodes a string
+- (NSString *)msidBase64UrlEncode
 {
-    NSData *decodedData = [self.class msidBase64UrlDecodeData:self];
+    NSData *decodedData = [self dataUsingEncoding:NSUTF8StringEncoding];
     
-    return [[NSString alloc] initWithData:decodedData encoding:NSUTF8StringEncoding];
+    return [decodedData msidBase64UrlEncodedString];
 }
 
-//Helper method to encode 3 bytes into a sequence of 4 bytes:
-//"static inline" is the way declare inline methods in LLVM
-static inline void Encode3bytesTo4bytes(char* output, int b0, int b1, int b2)
-{
-    output[0] = base64UrlEncodeTable[b0 >> 2];                                  // 6 MSB from byte 0
-    output[1] = base64UrlEncodeTable[((b0 << 4) & 0x30) | ((b1 >> 4) & 0x0f)];  // 2 LSB from byte 0 and 4 MSB from byte 1
-    output[2] = base64UrlEncodeTable[((b1 << 2) & 0x3c) | ((b2 >> 6) & 0x03)];  // 4 LSB from byte 1 and 2 MSB from byte 2
-    output[3] = base64UrlEncodeTable[b2 & 0x3f];
-}
 
-/// <summary>
-/// Base64 URL encode a set of bytes.
-/// </summary>
-/// <remarks>
-/// See RFC 4648, Section 5 plus switch characters 62 and 63 and no padding.
-/// For a good overview of Base64 encoding, see http://en.wikipedia.org/wiki/Base64
-/// </remarks>
-+ (NSString *)msidBase64UrlEncodeData:(NSData *)data
+- (NSString *)msidBase64UrlDecode
 {
-    if ( nil == data )
-        return nil;
-    
-    const byte *pbBytes = [data bytes];
-    int         cbBytes = (int)[data length];
-    
-    // Calculate encoded string size including padding. This may be more than is actually
-    // required since we will not pad and instead will terminate with null. The computation
-    // is the number of byte triples times 4 radix64 characters plus 1 for null termination.
-    int   encodedSize = 1 + ( cbBytes + 2 ) / 3 * 4;
-    char *pbEncoded = (char *)calloc( encodedSize, sizeof(char) );
-    
-    if(!pbEncoded){
-        return nil;
+    NSData *data = [self.class msidBase64UrlDecodeData:self];
+    if (!data) return nil;
+    
+    char lastByte;
+    [data getBytes:&lastByte range:NSMakeRange([data length] - 1, 1)];
+    
+    // We need to check for null terminated string data by looking at the last bit.
```

Comment:
if there's any Apple documentation referencing this behavior, would be great to add a link